kernel/src/context.rs

```diff
@@ -185,53 +186,84 @@ impl ReadCString for VmReader<'_, Fallible> {
         let max_len = self.remain();
         let mut buffer: Vec<u8> = Vec::with_capacity(max_len);
 
-        macro_rules! read_one_byte_at_a_time_while {
-            ($cond:expr) => {
-                while $cond {
-                    let byte = self.read_val::<u8>()?;
-                    buffer.push(byte);
-                    if byte == 0 {
-                        return Ok(CString::from_vec_with_nul(buffer)
-                            .expect("We provided 0 but no 0 is found"));
-                    }
-                }
-            };
+        if read_until_nul_byte(self, &mut buffer, max_len)? {
+            return Ok(CString::from_vec_with_nul(buffer).unwrap());
         }
 
-        // Handle the first few bytes to make `cur_addr` aligned with `size_of::<usize>`
-        read_one_byte_at_a_time_while!(
-            !is_addr_aligned(self.cursor() as usize) && buffer.len() < max_len
-        );
-
-        // Handle the rest of the bytes in bulk
-        let mut cloned_reader = self.clone();
-        while (buffer.len() + mem::size_of::<usize>()) <= max_len {
-            let Ok(word) = cloned_reader.read_val::<usize>() else {
-                break;
-            };
-
-            if has_zero(word) {
-                for byte in word.to_ne_bytes() {
-                    self.skip(1);
-                    buffer.push(byte);
-                    if byte == 0 {
-                        return Ok(CString::from_vec_with_nul(buffer)
-                            .expect("We provided 0 but no 0 is found"));
-                    }
+        // Maximum length exceeded before finding the null terminator
```

Comment:
Similarly, `nul terminator`. You can search for double validation at https://doc.rust-lang.org/std/ffi/struct.CString.html.